File: Chessko/Logic/UCIScoreParser.swift
import Foundation

// MARK: - UCI parser ocene
//
// Odvojen od `StockfishBridge`-a namerno: `StockfishBridge` je `actor` koji
// uvozi ChessKitEngine i ne moze u SwiftPM testni paket, a citanje ocene iz
// linije teksta je cista funkcija koja se mora testirati. Granica je tacno na
// mestu gde prestaje rad sa motorom, a pocinje rad sa stringom.
//
// FORMAT NIJE PRETPOSTAVLJEN NEGO IZMEREN. Prva verzija ovog parsera trazila je
// `<score> cp 34` i vracala `nil` na svakoj stvarnoj liniji. Pravi oblik je
// dobijen kompajliranjem i pokretanjem same biblioteke (ChessKitEngine 0.7.0,
// `EngineResponseInfo.swift:264` gradi `" <cp> \(cp)"`, a `cp` je `Double?`):
//
//     <info> <depth> 12 <score> <cp> 34.0
//     <info> <score> <cp> -250.0
//     <info> <score> <mate> 3
//     <info> <score> <cp> 34.0 <upperbound>
//
// Dakle: tag posle `<score>` nosi UGLASTE ZAGRADE, `cp` stize kao decimalan
// broj, a `<lowerbound>`/`<upperbound>` su zasebni tagovi bez vrednosti.

enum UCIScoreParser {

    /// Cita ocenu iz jedne UCI `info` linije, u tagovanom formatu koji vraca
    /// `ChessKitEngine`.
    ///
    /// Trazi token `<score>` pa cita tagove iza njega, umesto da se oslanja na
    /// fiksnu poziciju — redosled polja u `info` liniji nije garantovan i menja
    /// se od pretrage do pretrage (motor salje `<pv>` samo ponekad, `<seldepth>`
    /// samo na vecim dubinama).
    static func score(from line: String) -> EngineScore? {
        let tokens = line.split(separator: " ")
        guard let scoreIdx = tokens.firstIndex(of: "<score>") else { return nil }

        var i = tokens.index(after: scoreIdx)
        while i < tokens.endIndex {
            let next = tokens.index(after: i)
            switch tokens[i] {
            case "<cp>":
                // `cp` je u biblioteci `Double`, pa stize kao "34.0". `Int(...)`
                // bi na tome vratio nil — zato Double pa zaokruzivanje.
                guard next < tokens.endIndex, let v = Double(tokens[next]) else { return nil }
                return .cp(Int(v.rounded()))
            case "<mate>":
                guard next < tokens.endIndex, let v = Int(tokens[next]) else { return nil }
                return .mate(v)
            case "<lowerbound>", "<upperbound>":
                // Tagovi bez vrednosti. U praksi stoje IZA ocene, pa se dovde ne
                // stigne; preskacu se zbog redosleda na koji se ne oslanjamo.
                i = next
            default:
                // Izasli smo iz `<score>` bloka a da nismo nasli ni cp ni mate.
                return nil
            }
        }
        return nil
    }
}

File: ChesskoAndroid/app/src/main/java/com/veljkoni/chessko/logic/UCIScoreParser.kt
package com.veljkoni.chessko.logic

import com.veljkoni.chessko.models.EngineScore

// MARK: - Parser UCI ocene (Android)
//
// Namerno BEZ ijednog `android.*` uvoza -- isto pravilo kao `models/MoveAnalysis.kt`,
// `logic/PathProgress.kt` i `logic/PuzzleRating.kt`: cita se JVM testom bez emulatora.
//
// FORMAT NIJE PRETPOSTAVLJEN NEGO IZMEREN. iOS je isti parser prvo napisao za
// `<score> <cp> 34.0` (ChessKitEngine-ov tagovan oblik) i pao je na SVAKOJ
// stvarnoj liniji jer je pravi oblik bio drugaciji (vidi `Chessko/Logic/UCIScoreParser.swift`).
// Android ne koristi ChessKitEngine -- `logic/StockfishEngine.kt` je JNI most na
// PRAVI Stockfish (`cpp/stockfish/`), pa je izlaz obican UCI protokol, bez ijedne
// zagrade i bez decimale:
//
//     info depth 1 seldepth 3 multipv 1 score cp 445 nodes 52 ... pv d5c4
//     info depth 1 seldepth 4 multipv 1 score cp -18 nodes 58 ... pv f1d3
//     info depth 1 seldepth 3 multipv 1 score mate 1 nodes 31 ... pv d8h4
//     info depth 1 seldepth 3 multipv 1 score mate -1 nodes 2 ... pv a8d8 d1d8
//
// Uhvaceno preko `adb logcat` na emulatoru dok je aplikacija igrala stvarnu
// partiju na tezini "Stockfish Majstor" (pozitivan cp i pozitivan mat) i dok je
// motor pretrazivao poziciju iz vec isporucene `puzzles.sqlite` baze (negativan
// mat) -- vidi izvestaj Task-a 2. Potvrdjeno i iz izvora koji se kompajlira u
// ovu apk (`cpp/stockfish/uci.cpp:536`, `format_score`):
//
//     [](Score::InternalUnits units) { return "cp " + std::to_string(units.value); }
//     [](Score::Mate mate) { return "mate " + std::to_string(m); }
//
// -- prost `std::to_string` bez ikakvog posebnog tretmana znaka, pa je "-18"/"-1"
// obican predznak u broju, ne poseban tag.
//
// `lowerbound`/`upperbound` (bez vrednosti, odmah iza ocene -- `search.cpp:2166-2169`)
// oznacavaju NEKONACNU ocenu (aspiration window fail-high/fail-low usred iterative
// deepening-a) -- takva linija se namerno odbacuje, isto kao na iOS-u, jer bi
// analiza inace uzela ocenu koja ce se u sledecoj liniji iste dubine promeniti.
object UCIScoreParser {

    /**
     * Cita ocenu iz jedne UCI `info` linije native Stockfish-a.
     *
     * Trazi token "score" pa cita naredna dva tokena ("cp"/"mate" + broj), umesto
     * da se oslanja na fiksnu poziciju u liniji -- redosled polja iza ocene
     * (`nodes`, `nps`, `pv`, ...) nije nas posao, a `depth`/`seldepth`/`multipv`
     * ispred ocene ne menjaju gde "score" pocinje.
     */
    fun parse(line: String): EngineScore? {
        val tokens = line.trim().split(" ").filter { it.isNotEmpty() }
        val scoreIdx = tokens.indexOf("score")
        if (scoreIdx == -1 || scoreIdx + 2 >= tokens.size) return null

        val kind = tokens[scoreIdx + 1]
        val value = tokens[scoreIdx + 2].toIntOrNull() ?: return null

        // Nekonacna ocena (fail-high/fail-low usred pretrage) -- odbaci.
        val next = tokens.getOrNull(scoreIdx + 3)
        if (next == "lowerbound" || next == "upperbound") return null

        return when (kind) {
            "cp" -> EngineScore.Cp(value)
            "mate" -> EngineScore.Mate(value)
            else -> null
        }
    }
}
